src/main/java/io/github/intoto/helpers/IntotoHelper.java

```diff
@@ -0,0 +1,152 @@
+   * another convenience method that returns the serialized JSON representation for the envelope
+   *
+   * @param statement the Statement to add to the envelope
+   * @param signer the Signer that will be used to sign the payloads.
+   * @return will return a {@link IntotoEnvelope} instead of the JSON representation.
+   * @throws InvalidModelException thrown when the given statement is not valid
+   * @throws JsonProcessingException thrown when there are issues generating the JSON string
+   * @throws NoSuchAlgorithmException thrown when there are issues encrypting the payloads in the *
+   *     Envelope
+   * @throws SignatureException thrown when there are issues with the given key in the Signer
+   * @throws InvalidKeyException thrown when there are issues matching the key with the given *
+   *     algorithm
+   */
+  public static IntotoEnvelope produceIntotoEnvelope(Statement statement, Signer signer)
+      throws InvalidModelException, JsonProcessingException, NoSuchAlgorithmException,
+          SignatureException, InvalidKeyException {
+    // Get the Base64 encoded Statement to use as the payload
+    String jsonStatement = validateAndTransformToJson(statement, false);
+    String base64EncodedStatement = Base64.getEncoder().encodeToString(jsonStatement.getBytes());
+
+    IntotoEnvelope envelope = new IntotoEnvelope();
+    // Create the signed payload with the DSSEv1 format and sign it!
+    byte[] paeByteArray =
+        createPreAuthenticationEncoding(envelope.getPayloadType(), jsonStatement.getBytes());
+    byte[] signedDsseV1Payload = signer.sign(paeByteArray);
+    Signature signature = new Signature();
+    signature.setKeyId(signer.getKeyId());
+    // The sig contains the base64 encoded version of the signedDsseV1Payload
+    signature.setSig(Base64.getEncoder().encodeToString(signedDsseV1Payload));
+    // Let's complete the envelope
+    envelope.setPayload(base64EncodedStatement);
+    envelope.setSignatures(List.of(signature));
+    return envelope;
+  }
+
+  /**
+   * Generates the Pre-Authentication Encoding
+   *
+   * <pre>
+   * "DSSEv1" + SP + LEN(type) + SP + type + SP + LEN(body) + SP + body
+   *
+   * where:
+   * + = concatenation
+   * SP = ASCII space [0x20]
+   * "DSSEv1" = ASCII [0x44, 0x53, 0x53, 0x45, 0x76, 0x31]
+   * LEN(s) = ASCII decimal encoding of the byte length of s, with no leading zeros
+   * </pre>
+   *
+   * @param payloadType the type of payload. Fixed for in-toto Envelopes
+   * @param payload raw payload in bytes
+   * @return will return a Pre Authentication Encoding String.
+   */
+  public static byte[] createPreAuthenticationEncoding(String payloadType, byte[] payload) {
+    return String.format(
+            "DSSEv1 %d %s %d %s",
+            payloadType.length(),
+            payloadType,
+            payload.length,
+            new String(payload, StandardCharsets.UTF_8))
+        .getBytes(StandardCharsets.UTF_8);
```

Comment:
I'm also curious about what re-encoding as utf-8 results in the Java case. Does it take the bytes as is and assume that as long as they are in the allowed character set they are allowed? or does it perform a transformation on bytes it finds along the way?

(e.g., what if it finds a continuation byte followed  by something beyond 0x7F), does it add another continuation byte or?)

Reply:
So the `String.format` will automatically convert the payload into it's String representation. Calling out:

```Java
  return String.format(
            "DSSEv1 %d %s %d %s",
            payloadType.length(),
            payloadType,
            payload.length,
            new String(payload, StandardCharsets.UTF_8))
```

or

```Java
  return String.format(
            "DSSEv1 %d %s %d %s",
            payloadType.length(),
            payloadType,
            payload.length,
            payload)
```

Is basically the same thing, except that in the first case we are passing the correct encoding, instead of using the default.

In our case, the byte array (payload) is a String that contains the JSON that we ourselves generate. If this code lived in a separate DSSE implementation where the byte array can be anything (like a binary), then it would make sense to be extra careful on the conversion. In fact I was thinking of replacing all the byte shenanigans and just accept Strings in this method, but if we later want to move this to the DSSE and make it more generic it would make it harder.